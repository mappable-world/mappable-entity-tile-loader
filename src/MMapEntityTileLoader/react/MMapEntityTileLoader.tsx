import type TReact from 'react';
import type {MMapEntity} from '@mappable-world/mappable-types';
import type {CustomReactify} from '@mappable-world/mappable-types/reactify/reactify';
import type {
    GeojsonFeature,
    MMapEntityTileLoader as MMapEntityTileLoaderI,
    MMapEntityTileLoaderProps
} from '../MMapEntityTileLoader';
import throttle from 'lodash/throttle';

/**
 * Create reactified version of MMapEntityTileLoader module
 *
 * @example
 * ```jsx
 * <MMap location={LOCATION} ref={x => map = x}>
 *     <MMapDefaultSchemeLayer />
 *     <MMapDefaultFeaturesLayer />
 *     <MMapEntityTileLoader
 *          tileSize={TILE_SIZE}
 *          getFeatureId={useCallback((feature) => feature.id, [])}
 *          fetchTile={fetchTile}
 *          entity={useCallback(
 *              (feature) => (
 *                  <MMapFeature geometry={feature.geometry} properties={feature.properties} />
 *              ),
 *              []
 *          )}
 *          onFeatureAdd={useCallback((entity) => {
 *               setTotal((total) => total + entity.properties.area_sqkm);
 *          }, [])}
 *          onFeatureRemove={useCallback((entity) => {
 *              setTotal((total) => total - entity.properties.area_sqkm);
 *          }, [])}
 *     />
 * </MMap>
 * ```
 */

type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};

type MMapEntityTileLoaderReactifiedProps = Prettify<
    Omit<MMapEntityTileLoaderProps, 'entity'> & {
        /** Function that returns MMapEntity react component to render feature*/
        entity: (feature: GeojsonFeature) => TReact.ReactElement;

        /**
         * Every time a new tile is loaded and new features are added,
         * React will re-calls render for each added feature.
         * To avoid this, you can set renderDelay (in milliseconds).
         * After this time, React will render all features at once render call.
         * @default 300
         */
        renderDelay?: number;
    }
>;

type MMapEntityTileLoaderImperative = new (
    props: MMapEntityTileLoaderReactifiedProps
) => MMapEntity<MMapEntityTileLoaderReactifiedProps>;
type MMapEntityTileLoaderR = TReact.ForwardRefExoticComponent<
    MMapEntityTileLoaderReactifiedProps & TReact.RefAttributes<MMapEntity<MMapEntityTileLoaderReactifiedProps>>
>;

export const DEFAULT_THROTTLE_TIMOUT = 300;

export const MMapEntityTileLoaderReactifyOverride: CustomReactify<MMapEntityTileLoaderI, MMapEntityTileLoaderR> = (
    MMapEntityTileLoaderI,
    {reactify, React}
) => {
    const MMapEntityTileLoaderReactified = reactify.entity(
        MMapEntityTileLoaderI as unknown as MMapEntityTileLoaderImperative
    );

    const FeaturesList = React.memo(
        ({
            features,
            getFeatureId,
            entity
        }: {
            features: GeojsonFeature[];
            entity: MMapEntityTileLoaderReactifiedProps['entity'];
            getFeatureId: MMapEntityTileLoaderReactifiedProps['getFeatureId'];
        }) => {
            return (
                <>
                    {features.map((feature) => (
                        <React.Fragment key={getFeatureId(feature)}>{entity(feature)}</React.Fragment>
                    ))}
                </>
            );
        }
    );

    const MMapEntityTileLoader = React.forwardRef<
        MMapEntity<MMapEntityTileLoaderReactifiedProps>,
        MMapEntityTileLoaderReactifiedProps
    >((props, ref) => {
        const [features, setFeatures] = React.useState<GeojsonFeature[]>([]);

        const updateFeaturesList = React.useMemo(
            () => throttle(setFeatures, props.renderDelay ?? DEFAULT_THROTTLE_TIMOUT),
            [props.renderDelay ?? DEFAULT_THROTTLE_TIMOUT]
        );

        const onFeatureAdd = React.useCallback(
            (feature: GeojsonFeature): false => {
                props.onFeatureAdd?.(feature);

                if (!features.includes(feature)) {
                    features.push(feature);
                }

                updateFeaturesList([...features]);

                return false;
            },
            [features]
        );

        const onFeatureRemove = React.useCallback(
            (feature: GeojsonFeature): false => {
                props.onFeatureRemove?.(feature);
                const index = features.indexOf(feature);
                if (index !== -1) {
                    features.splice(index, 1);
                }

                updateFeaturesList([...features]);

                return false;
            },
            [features]
        );

        return (
            <>
                <MMapEntityTileLoaderReactified
                    {...props}
                    ref={ref}
                    onFeatureAdd={onFeatureAdd}
                    onFeatureRemove={onFeatureRemove}
                />
                <FeaturesList features={features} getFeatureId={props.getFeatureId} entity={props.entity} />
            </>
        );
    });

    return MMapEntityTileLoader;
};
