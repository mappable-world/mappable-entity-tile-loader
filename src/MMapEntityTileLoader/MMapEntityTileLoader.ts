import type {GenericFeature, LngLat, MMapEntity} from '@mappable-world/mappable-types';

import {MMapListener} from '@mappable-world/mappable-types';
import {MMapEntityTileLoaderReactifyOverride} from './react/MMapEntityTileLoader';
import {getTilesForViewport} from './helpers/getTilesForViewport';
import throttle from 'lodash/throttle';

export type GeojsonFeature = GenericFeature<LngLat>;

export interface MMapEntityTileLoaderProps {
    /** Tile size in pixels. World is 256x256 pixels on 0 zoom in Mappable */
    readonly tileSize: number;

    /**
     * Function for loading data by tile, should return an array of GeoJSON features
     */
    fetchTile: (args: {tx: number; ty: number; tz: number; signal: AbortSignal}) => Promise<GeojsonFeature[]>;

    /**
     * Function for getting the id of the feature.
     */
    getFeatureId: (feature: GeojsonFeature) => string;

    /**
     * Function for creating an [MMapEntity](https://mappable.world/docs/js-api/ref/index.html#class-mmapentity) from a feature.
     */
    entity: (feature: GeojsonFeature) => MMapEntity<unknown>;

    /**
     * Function is called when a feature is added to the map.
     * If the function returns `false`, the feature will not be added to the map.
     * In this case, you should add the feature to the map yourself.
     */
    onFeatureAdd: (feature: GeojsonFeature) => void | false;

    /**
     * Function is called when a feature is removed from the map.
     * If the function returns `false`, the feature will not be removed from the map.
     * In this case, you should remove the feature from the map yourself.
     */
    onFeatureRemove: (feature: GeojsonFeature) => void | false;

    /**
     * By default, when changing tiles, old features are immediately deleted.
     * But the same points may appear in the new tile, then there was no point in deleting them.
     * Set the delay for applying deletion operations.
     * @default 0
     */
    removalDelay?: number;
}

interface Tile {
    tx: number;
    ty: number;
    tz: number;
    abortController: AbortController;
    entities: Map<string, SharedFeature>;
    promise: Promise<void>;
}

interface SharedFeature {
    id: string;
    refcount: number;
    feature: GeojsonFeature;
    entity?: MMapEntity<unknown>;
}

export class MMapEntityTileLoader extends mappable.MMapComplexEntity<MMapEntityTileLoaderProps> {
    static [mappable.overrideKeyReactify] = MMapEntityTileLoaderReactifyOverride;

    private _tiles = new Map<string, Tile>();
    private _features = new Map<string, SharedFeature>();
    private _listener: MMapListener;
    private _requestDeleteFeatures: Function;

    constructor(props: MMapEntityTileLoaderProps) {
        super({removalDelay: 0, ...props}, {container: true});

        this._listener = new mappable.MMapListener({
            onUpdate: ({mapInAction}) => {
                if (!mapInAction) {
                    this._reconcileTiles();
                }
            }
        });

        this._addDirectChild(this._listener);

        this._requestDeleteFeatures = this._props.removalDelay
            ? throttle(() => this._deleteFeatures(), this._props.removalDelay)
            : () => this._deleteFeatures();
    }

    protected override _onUpdate({removalDelay}: Partial<MMapEntityTileLoaderProps>) {
        if (removalDelay !== undefined) {
            this._requestDeleteFeatures = removalDelay
                ? throttle(() => this._deleteFeatures(), removalDelay)
                : () => this._deleteFeatures();
        }
    }

    protected _onDetach() {
        for (const child of this.children) {
            this.removeChild(child);
        }

        this._tiles.clear();
        this._features.clear();
    }

    protected _onAttach() {
        this._reconcileTiles();
    }

    private _markedForDeletion: Set<SharedFeature> = new Set();

    private _reconcileTiles() {
        const {projection, zoom, center, size} = this.root!;
        const tilesForViewport = getTilesForViewport({
            projection,
            zoom,
            center,
            size,
            tileSize: this._props.tileSize
        });

        const tiles = new Map();
        for (const tile of tilesForViewport) {
            tiles.set(`${tile.tx}_${tile.ty}_${tile.tz}`, tile);
        }

        // Add new
        for (const [key, description] of tiles.entries()) {
            if (this._tiles.has(key)) continue;

            const tile: Tile = {
                ...description,
                abortController: new AbortController(),
                entities: new Map(),
                promise: Promise.resolve()
            };

            tile.promise = this._fetchTile(tile);

            this._tiles.set(key, tile);
        }

        // Remove old tiles and features
        for (const [key, tile] of this._tiles.entries()) {
            if (tiles.has(key)) continue;
            this._tiles.delete(key);

            for (const sharedEntity of tile.entities.values()) {
                sharedEntity.refcount--;
                if (sharedEntity.refcount === 0) {
                    this._markedForDeletion.add(sharedEntity);
                }
            }

            tile.abortController.abort();
        }

        this._requestDeleteFeatures();
    }

    private async _fetchTile(tile: Tile): Promise<void> {
        const [ok, featuresOrError] = await this._props
            .fetchTile({tx: tile.tx, ty: tile.ty, tz: tile.tz, signal: tile.abortController.signal})
            .then(
                (data) => [true, data],
                (error) => [false, error]
            );

        if (!ok || !Array.isArray(featuresOrError)) {
            if (
                !featuresOrError ||
                (featuresOrError instanceof DOMException && featuresOrError.name === 'AbortError')
            ) {
                return;
            }
            throw featuresOrError;
        }

        for (const feature of featuresOrError) {
            const id = this._props.getFeatureId(feature);
            if (tile.entities.has(id)) continue;

            const existingFeature = this._features.get(id);
            const sharedFeature = existingFeature ?? {
                id,
                feature,
                refcount: 0
            };

            sharedFeature.refcount++;
            tile.entities.set(id, sharedFeature);

            if (!existingFeature) {
                this._features.set(id, sharedFeature);
                this.__addFeature(sharedFeature);
            }
        }
    }

    private _deleteFeatures() {
        this._markedForDeletion.forEach((sharedEntity) => {
            if (sharedEntity.refcount !== 0) {
                return;
            }

            this._features.delete(sharedEntity.id);
            this.__removeFeature?.(sharedEntity);
        });

        this._markedForDeletion.clear();
    }

    private __addFeature(sharedEntity: SharedFeature): void {
        if (this._props.onFeatureAdd?.(sharedEntity.feature) === false) {
            return;
        }

        if (!sharedEntity.entity) {
            sharedEntity.entity = this._props.entity(sharedEntity.feature);
        }

        this._addDirectChild(sharedEntity.entity);
    }

    private __removeFeature(sharedEntity: SharedFeature): void {
        if (this._props.onFeatureRemove?.(sharedEntity.feature) === false) {
            return;
        }

        this._removeDirectChild(sharedEntity.entity);
    }
}
