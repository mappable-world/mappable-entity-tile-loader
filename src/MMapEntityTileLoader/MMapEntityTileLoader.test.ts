import type {LngLatBounds} from '@mappable-world/mappable-types/common/types';
import {MMap} from '@mappable-world/mappable-types';
import {GeojsonFeature, MMapEntityTileLoader} from './MMapEntityTileLoader';

function nextTick(): Promise<void> {
    return new Promise((resolve) => process.nextTick(() => resolve()));
}

describe('MMap smoke test', () => {
    const BOUNDS: LngLatBounds = [
        [53.20890963521473, 25.52765018907181],
        [57.444403818421854, 24.71096299361919]
    ];

    const FEATURES: GeojsonFeature[] = [
        // prettier-ignore
        {"type":"Feature","properties":{"scalerank":5,"area_sqkm":306.708,"featureclass":"Urban area"},"geometry":{"type":"Polygon","coordinates":[[[57.27808556501731,23.90505280201225],[57.27808556501731,23.90505280201225],[57.27808556501731,23.90505280201225],[57.61129520043531,23.803663641855593],[57.70147057478462,23.78490509693364],[57.701987339107546,23.77741201425131],[57.72234785343056,23.738137925709495],[57.63558312361249,23.720981350188595],[57.5452527199663,23.724004421477673],[57.49605675642445,23.768446153248675],[57.44365685408047,23.812913723235837],[57.36929446801247,23.830897121673402],[57.29126305525173,23.850999253834942],[57.27839562361106,23.89569936776742],[57.27808556501731,23.90505280201225]]]},"id":'1436'},
        // prettier-ignore
        {"type":"Feature","properties":{"scalerank":5,"area_sqkm":196.842,"featureclass":"Urban area"},"geometry":{"type":"Polygon","coordinates":[[[55.94121626162655,25.793981431577095],[55.94121626162655,25.793981431577095],[55.94121626162655,25.793981431577095],[56.02524214053315,25.9159894882182],[56.047359653554054,25.97722606048407],[56.04927168154887,25.976321722918968],[56.08833906436152,25.906610215757226],[56.06115726097599,25.81713247324386],[56.016663852772695,25.77157969817857],[55.95852786644434,25.744578762306062],[55.89264041527218,25.73791250254041],[55.87010949079294,25.743700262957105],[55.94121626162655,25.793981431577095]]]},"id":'1429'},
        // prettier-ignore
        {"type":"Feature","properties":{"scalerank":3,"area_sqkm":269.775,"featureclass":"Urban area"},"geometry":{"type":"Polygon","coordinates":[[[57.99638797387436,23.73444306080063],[57.99638797387436,23.73444306080063],[57.99638797387436,23.73444306080063],[58.120411411374846,23.71653717701149],[58.27766279483902,23.645120347584125],[58.28045332218278,23.62724030201113],[58.2766292661932,23.616336574797543],[58.270686476479625,23.612874253833994],[58.26128136580249,23.608740139250642],[58.209243198484586,23.589878241464106],[58.154879591713524,23.58429718677658],[58.114727003822736,23.61403697356056],[58.07865685408302,23.647807522063303],[58.01824710473383,23.657264309172717],[57.977060988197195,23.685893052662408],[57.98677615746806,23.721704820240674],[57.99638797387436,23.73444306080063]]]},"id":'1437'},
        // prettier-ignore
        {"type":"Feature","properties":{"scalerank":5,"area_sqkm":171.273,"featureclass":"Urban area"},"geometry":{"type":"Polygon","coordinates":[[[56.64060509626478,24.47028961842075],[56.64060509626478,24.47028961842075],[56.64060509626478,24.47028961842075],[56.774136997306954,24.334587307222293],[56.88736006045846,24.18371796314618],[56.861108432854195,24.170721340424763],[56.81826867048423,24.213380235281704],[56.783645460848675,24.266942857352234],[56.74736860537979,24.314175116467013],[56.70804284040565,24.358952745047915],[56.66148237491069,24.39631480559494],[56.61388838076988,24.431093044527373],[56.59156416201978,24.460109361259256],[56.58965213402499,24.48672272388957],[56.61435346866048,24.513129380790716],[56.64060509626478,24.47028961842075]]]},"id":'1434'}
    ];

    const LOCATION = {bounds: BOUNDS};

    let container: HTMLElement, map: MMap, total: number, tileLoader: MMapEntityTileLoader;

    beforeEach(() => {
        container = document.createElement('div');
        Object.assign(container.style, {width: `643px`, height: `856px`});
        document.body.appendChild(container);
        map = new mappable.MMap(container, {location: LOCATION}, [
            new mappable.MMapDefaultSchemeLayer({}),
            new mappable.MMapDefaultFeaturesLayer({})
        ]);

        total = 0;
        tileLoader = new MMapEntityTileLoader({
            delayDeletion: 0,
            tileSize: 256, // World is 256x256 pixels on 0 zoom in 3.0.
            getFeatureId: (feature) => feature.id,
            fetchTile: ({tx, ty}) => {
                if (tx === 167) {
                    switch (ty) {
                        case 107:
                            return Promise.resolve([FEATURES[0]]);
                        case 108:
                            return Promise.resolve([FEATURES[1]]);
                        case 109:
                            return Promise.resolve([FEATURES[2]]);
                    }
                }

                if (tx === 337) {
                    switch (ty) {
                        case 219:
                            return Promise.resolve([FEATURES[3]]);
                    }
                }

                return Promise.resolve([]);
            },

            entity: (feature) =>
                new mappable.MMapFeature({
                    id: feature.id as string,
                    geometry: feature.geometry,
                    properties: feature.properties
                }),
            onFeatureAdd: (feature) => {
                total += feature.properties.area_sqkm as number;
            },
            onFeatureRemove: (feature) => {
                total -= feature.properties.area_sqkm as number;
            }
        });

        map.addChild(tileLoader);
    });

    afterEach(() => {
        map.removeChild(tileLoader);
        map.destroy();
        container.remove();
    });

    it('should add several polygons on map', async () => {
        await nextTick();

        const tree = domToJson(map.container);
        expect(tree).toMatchSnapshot();

        // @ts-ignore We do not add public children
        expect(tileLoader.children.length).toEqual(0);
        expect(document.querySelectorAll('path').length).toEqual(3);
        expect(Math.round(total)).toEqual(773);
    });

    describe('Move map', () => {
        it('should load and render another features', async () => {
            map.setLocation({center: [57.859760289989055, 23.678781782281117], zoom: 9});
            await nextTick();

            const tree = domToJson(map.container);
            expect(tree).toMatchSnapshot();

            // @ts-ignore We do not add public children
            expect(tileLoader.children.length).toEqual(0);
            expect(document.querySelectorAll('path').length).toEqual(1);
            expect(total).toEqual(FEATURES[3].properties.area_sqkm);
        });
    });
});
