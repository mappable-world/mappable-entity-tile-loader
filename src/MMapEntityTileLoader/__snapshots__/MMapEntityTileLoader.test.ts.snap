// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MMap smoke test Move map should load and render another tiles 1`] = `
{
  "attributes": {
    "class": "mappable--map-container",
  },
  "children": [
    {
      "attributes": {
        "class": "mappable--map",
        "style": "width: 643px; height: 856px;",
      },
      "children": [
        {
          "attributes": {
            "class": "mappable--main-engine-container",
          },
          "children": [
            {
              "attributes": {},
              "children": [
                {
                  "attributes": {
                    "class": "mappable--layer mappable--tile-layer",
                    "style": "perspective: 1597.3177456394797px; opacity: 1;",
                  },
                  "children": [
                    {
                      "attributes": {
                        "class": "mappable--tile-layer__container",
                        "data-z": "8",
                        "style": "transform: 
            rotateX(0.00000rad)
            translate3d(0px, 0px, 798.6588728197398px)
            rotateZ(0.00000rad) translate3d(-461px, -288px, 0px)
        ; z-index: -1;",
                      },
                      "nodeName": "mappable",
                    },
                    {
                      "attributes": {
                        "class": "mappable--tile-layer__container",
                        "data-z": "9",
                        "style": "transform: 
            rotateX(0.00000rad)
            translate3d(0px, 0px, 0px)
            rotateZ(0.00000rad) translate3d(0px, 0px, 0px)
        ; z-index: 9;",
                      },
                      "nodeName": "mappable",
                    },
                  ],
                  "nodeName": "mappable",
                },
              ],
              "nodeName": "mappable",
            },
          ],
          "nodeName": "mappable",
        },
        {
          "attributes": {
            "class": "mappable--top-engine-container",
          },
          "children": [
            {
              "attributes": {},
              "children": [
                {
                  "attributes": {
                    "class": "mappable--layer mappable--graphics-layer",
                    "style": "perspective: 1597.3177456394797px; opacity: 1;",
                  },
                  "children": [
                    {
                      "attributes": {
                        "height": "1712",
                        "style": "transform: translate3d(-321.5px, -428px, 0px) rotateX(0rad) rotateZ(0rad) translate3d(0px, 0px, 0px) scale(1); will-change: transform;",
                        "viewBox": "0 0 1286 1712",
                        "width": "1286",
                        "xlink": "http://www.w3.org/1999/xlink",
                      },
                      "children": [
                        {
                          "attributes": {
                            "opacity": "1",
                            "transform": "translate(643, 856)",
                          },
                          "children": [
                            {
                              "attributes": {
                                "fill": "rgb(0,102,255)",
                                "fill-opacity": "0.6",
                                "fill-rule": "evenodd",
                                "id": "mappable--feature-0-__auto_3",
                                "stroke-dasharray": "",
                                "stroke-linecap": "round",
                                "stroke-linejoin": "round",
                                "stroke-width": "1.000",
                                "style": "cursor: inherit;",
                              },
                              "children": [
                                {
                                  "attributes": {
                                    "d": "M-443.880859866189,-315.638009292079 L-395.2633783845231,-261.38352329821464 L-354.04011912509304,-201.13334261542514 L-363.5980450510542,-195.9464370125079 L-379.1955265325996,-212.97342900237527 L-419.32752653276475,-271.1206608729335 L-453.6081191254889,-299.9610257774184 L-461.73611912552224,-311.56586507589054 L-462.43226727367073,-322.21201310191645 L-453.43878579216107,-332.777698855416 L-443.880859866189,-315.638009292079 Z",
                                    "data-feature-id": "__auto_3",
                                    "stroke": "rgb(0,102,255)",
                                    "stroke-opacity": "1",
                                  },
                                  "nodeName": "path",
                                },
                              ],
                              "nodeName": "g",
                            },
                          ],
                          "nodeName": "g",
                        },
                      ],
                      "nodeName": "svg",
                    },
                  ],
                  "nodeName": "mappable",
                },
                {
                  "attributes": {
                    "class": "mappable--layer mappable--marker-layer",
                    "style": "opacity: 1; transform-origin: 321.5px 428px;",
                  },
                  "nodeName": "mappable",
                },
              ],
              "nodeName": "mappable",
            },
          ],
          "nodeName": "mappable",
        },
      ],
      "nodeName": "mappable",
    },
    {
      "attributes": {
        "class": "mappable--map-copyrights mappable--map-copyrights_bottom mappable--map-copyrights_right",
      },
      "children": [
        {
          "attributes": {
            "class": "mappable--map-copyrights__container",
          },
          "children": [
            {
              "attributes": {
                "class": "mappable--map-copyrights__text",
              },
              "nodeName": "span",
            },
            {
              "attributes": {
                "class": "mappable--map-copyrights__user-agreements",
                "href": "https://mappable.world/legal/tos",
                "target": "_blank",
              },
              "children": [
                "User Agreement",
              ],
              "nodeName": "a",
            },
          ],
          "nodeName": "div",
        },
        {
          "attributes": {
            "class": "mappable--map-copyrights__logo mappable--icon-logo-en-black",
            "style": "width: 80px; height: 20px;",
          },
          "nodeName": "span",
        },
      ],
      "nodeName": "div",
    },
  ],
  "nodeName": "mappable",
}
`;
